### PR TITLE
Create crud endpoints for destination

### PR DESCRIPTION
**Pull request by shizuko-akamoto**
SO, turns out project_4 is quite messy, and I don't wanna overwhelm you guys with another 50+ file changes in a single PR, so I decided to clean up and split it up to smaller merges. 

This PR is really just defining server endpoints for Destination API.

Defined some destination routes that we are/will probably use in frontend. So far, these three endpoints:
- `createDestination`
- `getDestinationByTrecipeId`: this one is implemented by first finding the trecipe document with given uuid, then using mongoose's `populate` method to fill in its `destinations` field. Since each `destWithStatus` object in `destinations` field has a virtual field acting as foreign key to actual destination mode, MongoDB is able to populate the field with actual destinations document.
- `getDestinationById`

Other notable changes:
- We currently have just too many versions of interfaces representing the same model! One on client, one on server, one for OpenAPI documentation..... Everytime there's a change in model we end up having to duplicate the updates. So, I decided to move the files defining models into a `shared` folder, which is supposed to be both accessible from the server and the client.
I've already switched the server to use the models in the shared folder, but I haven't touched the client. Will do that in next PR (which is to connect the endpoints up)
- Flipped the order of some then and catch blocks in server services. The reason being, if in a then block, we return a promise rejection and there is a catch block following, then the rejection just gets caught immediately by that following catch block (rather than the catch block of the calling method).

### CODE REVIEW

File: packages/server/server/api/destinations/destination.service.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    public getDestinationsByTrecipeId(trecipeUuid: string): Promise<Array<DestWithStatus>> {
+        const populateField = 'destinations.destination';
+        return trecipeModel
+            .findOne({ uuid: trecipeUuid })
+            .populate(populateField)
+            .exec()
+            .catch((err) => {
+                logger.info(`failed to get destinations for trecipe with uuid: ${trecipeUuid}`);
```
[the remark this answers is not in the document]
Comment (shizuko-akamoto) [reply packages/server/server/api/destinations/destination.service.ts:47]: Oops missed that. Thankss

File: packages/server/server/api/destinations/destination.service.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    public updateDestinationById(uuid: string, destData: Destination): Promise<Destination> {
+        return destinationModel
+            .findOneAndUpdate({ uuid: uuid }, destData)
+            .exec()
+            .catch((err) =>
+                Promise.reject(
+                    new InternalServerError({
+                        message: `Failed to get destination: ${err.toString()}`,
```
[the remark this answers is not in the document]
Comment (shizuko-akamoto) [reply packages/server/server/api/destinations/destination.service.ts:94]: hahaha good catch

File: packages/server/server/common/api.yml [hunk 3]
```diff
@@ -326,6 +324,50 @@ components:
         - name
         - description
         - isPrivate
+    CreateNewDestinationDTO:
+      title: CreateNewDestinationDTO
+      type: object
+      properties:
```
[the remark this answers is not in the document]
Comment (shizuko-akamoto) [reply packages/server/server/common/api.yml:330]: I think there are some libraries for building typescript models out of OpenAPI doc, or the other way around, autogenerating the doc based on typescript model...., but there's definitely a bit of learning curve in adding those in. Hopefully we don't have too many endpoints to declare further 😅

[the remark this answers is not in the document]
Comment (knox153) [reply packages/server/server/common/api.yml:330]: haha alright thanks for adding most of them!!

**Pull request merged**